File: src/main/java/no/nav/personopplysninger/features/endreopplysninger/domain/Endring.java
package no.nav.personopplysninger.features.endreopplysninger.domain;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;


@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Endring<T> {
    private String endringstype;
    private String ident;
    private String lineage;
    private String opplysningsId;
    private Status status;
    private String statusType = "OK";
    private ValidationError validationError;

    public String getEndringstype() {
        return endringstype;
    }

    public String getIdent() {
        return ident;
    }

    public String getLineage() {
        return lineage;
    }

    public String getOpplysningsId() {
        return opplysningsId;
    }

    public Status getStatus() {
        return status;
    }

    public String getStatusType() {
        return statusType;
    }

    public void setStatusType(String statusType) {
        this.statusType = statusType;
    }

    public ValidationError getValidationError() {
        return validationError;
    }

    public void setValidationError(ValidationError validationError) {
        this.validationError = validationError;
    }

    public void createValidationErrorIfTpsHasError() {
        if (hasTpsError()) {
            ValidationError validationError = new ValidationError();
            validationError.setMessage(getTpsBeskrivelse());
            setValidationError(validationError);
        }
    }

    @JsonIgnore
    public boolean isPending() {
        return "PENDING".equals(status.getStatusType());
    }

    @JsonIgnore
    public boolean isDoneWithoutTpsError() {
        if (!"DONE".equals(status.getStatusType())) {
            return false;
        }
        return !hasTpsError();
    }

    @JsonIgnore
    private boolean hasTpsError() {
        for (Substatus substatus: status.getSubstatus()) {
            if ("TPS".equalsIgnoreCase(substatus.getDomene())) {
                return "ERROR".equals(substatus.getStatus());
            }
        }
        return false;
    }

    @JsonIgnore
    private String getTpsBeskrivelse() {
        for (Substatus substatus: status.getSubstatus()) {
            if ("TPS".equalsIgnoreCase(substatus.getDomene())) {
                return substatus.getBeskrivelse();
            }
        }
        return null;
    }
}

File: src/main/kotlin/no/nav/personopplysninger/features/endreopplysninger/domain/Endring.kt
package no.nav.personopplysninger.features.endreopplysninger.domain

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
open class Endring<T> {
    val endringstype: String? = null
    val ident: String? = null
    val lineage: String? = null
    val opplysningsId: String? = null
    val opplysningstype: String? = null
    var status: Status = Status()
    var statusType = "OK"
    var error: Error? = null

    val confirmedOk: Boolean
        @JsonIgnore
        get() = "DONE" == status.statusType && !hasTpsError()

    val errorMessage: String
        @JsonIgnore
        get() =
            if (hasTpsError()) "TPS rapporterer feil på oppdatering."
            else if (!isDone) "Endring har status $statusType"
            else ""

    val isPending: Boolean
        @JsonIgnore
        get() = "PENDING" == status.statusType

    private val isDone: Boolean
        @JsonIgnore
        get() = "DONE" == status.statusType

    val hasTpsError: Boolean
        @JsonIgnore
        get() = hasTpsError()

    private val tpsBeskrivelse: String?
        @JsonIgnore
        get() {
            for (substatus in status.substatus) {
                if ("TPS".equals(substatus.domene, ignoreCase = true)) {
                    return substatus.beskrivelse
                }
            }
            return null
        }

    fun createValidationErrorIfTpsHasError() {
        if (hasTpsError()) {
            val validationError = Error()
            validationError.message = tpsBeskrivelse
            statusType = "ERROR"
            error = validationError
        }
    }

    @JsonIgnore
    private fun hasTpsError(): Boolean {
        for (substatus in status.substatus) {
            if ("TPS".equals(substatus.domene, ignoreCase = true)) {
                return "ERROR" == substatus.status
            }
        }
        return false
    }
}
